probe-rs/src/probe.rs

```diff
@@ -722,6 +722,11 @@ impl DebugProbeInfo {
     pub fn open(&self, lister: &Lister) -> Result<Probe, DebugProbeError> {
         lister.open(DebugProbeSelector::from(self))
     }
+
+    /// Returns whether this info was returned by a particular probe factory.
+    pub fn is_probe_type<F: ProbeFactory>(&self, factory: &F) -> bool {
+        self.probe_type == factory as &dyn ProbeFactory
```

Comment:
I don't think this actually works, see https://github.com/probe-rs/probe-rs/issues/2236#issuecomment-1965974125

Reply:
Why not? the test passes. At least give me a repro of your issue other than "it should look like Y"

Reply:
It doesn't work in practice, I'm not entirely sure why. It's reproducible if you try to use this with an actual probe.

I guess it's because what the code does is just comparing pointers,
and that works fine for your test case because it's the same object.

But when you use the `AllProbesLister`, you cannot get the internal `CmsisDapFactory` which is used here:
https://github.com/probe-rs/probe-rs/blob/5874daab70b8829f5f30459a173aaa006c97af83/probe-rs/src/probe/list.rs#L78

So you have to create your own, and then the pointer comparison fails.

Reply:
Thanks. We're not just comparing pointers, we're comparing type_id, _then_ pointers. We compare pointers because if the factory is not a ZST, you might be interested in different instances of it. The mistake is that I assumed ZST references are always (dangling pointer, metadata pointer) which seems to not be the case. Will look into it later.

Reply:
> Thanks. We're not just comparing pointers, we're comparing type_id, _then_ pointers. We compare pointers because if the factory is not a ZST, you might be interested in different instances of it. The mistake is that I assumed ZST references are always (dangling pointer, metadata pointer) which seems to not be the case. Will look into it later.

Having different instances of the factory seems annoying with the `&'static` lifetime which is currently required. 

As I said in an earlier comment, I'm fine with just removing the `probe_type`, for my usage switching to just listing cmsis dap probes is probably the better thing to do.

Reply:
> I'm fine with just removing the probe_type

It's already used to print the probe type in `probe-rs list`, so how about we try and make it bear more work? :)